### PR TITLE
Fixes #1178: Impersonate-Group may be specified multiple times

### LINKED ISSUE #1178: Impersonate-Group may be specified multiple times
According to https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation, the `Impersonate-Group` HTTP header can be specified multiple times, however `RequestConfig.setImpersonatorGroup` only accepts a single group.

### PR DESCRIPTION
**Pull request by gastaldi**
#1178 According to https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation, the `Impersonate-Group` header may be specified multiple times

### CODE REVIEW

**Comment by rohanKanojia** [comment]
ok to test

**Comment by rohanKanojia** [comment]
@bartoszmajsak : Hey, we already run our integration tests on single node openshift cluster on CircleCI

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Config.java [hunk 1]
```diff
@@ -485,7 +490,7 @@ private static boolean loadFromKubeconfig(Config config, String context, String
           }
 
           config.getErrorMessages().put(401, "Unauthorized! Token may have expired! Please log-in again.");
-          config.getErrorMessages().put(403, "Forbidden! User "+currentContext.getUser()+ " doesn't have permission.");
+          config.getErrorMessages().put(403, "Forbidden! User " + currentContext.getUser() + " doesn't have permission.");
```
Comment (rohanKanojia) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Config.java]: Minor nitpick, Can we get rid of this hardcoding and use `HttpUrlConnection` statuses here and at other places also? ;-)

**Comment by bartoszmajsak** [comment]
> @bartoszmajsak : Hey, we already run our integration tests on single node openshift cluster on CircleCI

@rohanKanojia I know you run them :) My ask was to create test cases covering this use case.

**Comment by rohanKanojia** [comment]
@bartoszmajsak :okay :smile:

**Comment by rohanKanojia** [comment]
@gastaldi @bartoszmajsak : Not sure why tests failed on `ci.fabric8.io`. They passed on my system locally.

**Comment by gastaldi** [comment]
@bartoszmajsak Added `UserImpersonationIT`, please review

**Comment by bartoszmajsak** [comment]
I'm reviewing it right now

File: kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    // Reset original authentication
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(null);
+    requestConfig.setImpersonateGroups(null);
+    client.rbac().kubernetesClusterRoles().inNamespace(currentNamespace).delete(impersonatorRole);
+    client.rbac().kubernetesClusterRoleBindings().inNamespace(currentNamespace).delete(impersonatorRoleBinding);
+    client.projects().delete(new ProjectBuilder().withNewMetadata().withName(NEW_PROJECT).endMetadata().build());
+    client.serviceAccounts().inNamespace(currentNamespace).delete();
+    // Wait for resources to get destroyed
+    Thread.sleep(3000);
```
Comment (bartoszmajsak) [inline kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: How about using Awaitility https://github.com/fabric8io/kubernetes-client/pull/1121/files#diff-e892e898797ea5223e863136b7348d7cR140 instead?

File: kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    // Create a project
+    ProjectRequest projectRequest = client.projectrequests().createNew()
+      .withNewMetadata()
+      .withName(NEW_PROJECT)
+      .endMetadata()
+      .done();
+
+    // Grab the requester annotation
+    String requester = projectRequest.getMetadata().getAnnotations().get("openshift.io/requester");
+    assertThat(requester).isEqualTo(SERVICE_ACCOUNT);
```
Comment (bartoszmajsak) [inline kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: I don't grok how in fact we are testing that we can impersonate groups set in the request? Is that a matter of better test description or we should make different assertions here?

Comment (gastaldi) [reply kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: There are only two groups in this case: https://docs.openshift.com/container-platform/3.10/architecture/additional_concepts/authentication.html#authentication-impersonation

File: kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java (hunk 2 above)
Comment (gastaldi) [reply kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: Not sure if that's really needed. I'd rather remove the Thread.sleep now that I see that it doesn't matter when the resources get discarded

File: kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java (hunk 3 above)
Comment (bartoszmajsak) [reply kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: Ok, so what you are testing here is if you can create a project impersonating other user with those groups. Then maybe the test name can be more descriptive. WDYT?

File: kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  public void cleanup() {
+    // Reset original authentication
+    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
+    requestConfig.setImpersonateUsername(null);
+    requestConfig.setImpersonateGroups(null);
+
+    client.rbac().kubernetesClusterRoles().inNamespace(currentNamespace).delete(impersonatorRole);
+    client.rbac().kubernetesClusterRoleBindings().inNamespace(currentNamespace).delete(impersonatorRoleBinding);
+    client.projects().delete(new ProjectBuilder().withNewMetadata().withName(NEW_PROJECT).endMetadata().build());
+    client.serviceAccounts().inNamespace(currentNamespace).delete();
```
Comment (bartoszmajsak) [inline kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: So we don't have to wait anymore?

File: kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java (hunk 3 above)
Comment (gastaldi) [reply kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: Any suggestions on the test name? `testCreateProjectImpersonatingServiceAccount` seemed pretty descriptive to me :smiley:

Comment (bartoszmajsak) [reply kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: groups sth sth

Comment (bartoszmajsak) [reply kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: "I should be able to create a project impersonating user while using service account and groups" maybe

File: kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        }
+      }
+    );
+    // Delete ServiceAccounts
+    client.serviceAccounts().inNamespace(currentNamespace).delete();
+    await().atMost(1, TimeUnit.MINUTES).until(
+      new Callable<Boolean>() {
+        @Override
+        public Boolean call() {
+          return client.serviceAccounts().list().getItems().isEmpty();
```
Comment (bartoszmajsak) [inline kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: How about making those anonymous classes methods named after the comments? 

`await().atMost(1, TimeUnit.MINUTES).until(serviceAccountIsDeleted())`?

Also - do we have to wait for all the steps? What is we e.g. make project name "randomized"?

Comment (gastaldi) [reply kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: Done.  Let' s keep it waiting for all the steps for now, as it's safer this way

File: kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java (hunk 3 above)
Comment (gastaldi) [reply kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java]: Done

**Comment by gastaldi** [comment]
[test]

**Comment by gastaldi** [comment]
Retest this please

**Comment by gastaldi** [comment]
I have no idea why the PR failed the tests, doesn't seem to be related to my changes

**Comment by gastaldi** [comment]
@bartoszmajsak I think it's all good now, can you review again?

**Comment by bartoszmajsak** [comment]
Awesome, many thanks @gastaldi 🥇

**Comment by rohanKanojia** [comment]
[merge]

**Pull request merged**